File: DSA/VivaswanSingh_590023022/Day_86/q2.c
#include <stdlib.h>

typedef struct
{
    int value;
    int weight;
    double ratio;
} Item;

int compare(const void *a, const void *b)
{
    Item *x = (Item *)a;
    Item *y = (Item *)b;

    if (x->ratio < y->ratio)
        return 1;

    if (x->ratio > y->ratio)
        return -1;

    return 0;
}

double fractionalKnapsack(int n, int value[], int weight[], int capacity)
{
    Item items[n];

    for (int i = 0; i < n; i++)
    {
        items[i].value = value[i];
        items[i].weight = weight[i];
        items[i].ratio = (double)value[i] / weight[i];
    }

    qsort(items, n, sizeof(Item), compare);

    double totalValue = 0.0;
    int remaining = capacity;

    for (int i = 0; i < n; i++)
    {
        if (remaining == 0)
            break;

        if (items[i].weight <= remaining)
        {
            totalValue += items[i].value;
            remaining -= items[i].weight;
        }
        else
        {
            totalValue += items[i].ratio * remaining;
            remaining = 0;
        }
    }

    return totalValue;
}

File: DSA/vaishnavi_joshi_590016947/Day86/Q2.java
import java.util.*;

class Solution {

    static class Item {
        long value;
        long weight;
        double ratio;

        Item(long value, long weight) {
            this.value = value;
            this.weight = weight;
            this.ratio = (double) value / weight;
        }
    }

    public static double fractionalKnapsack(
            int n, long[] value, long[] weight, long capacity) {

        Item[] items = new Item[n];

        // Create items
        for (int i = 0; i < n; i++) {
            items[i] = new Item(value[i], weight[i]);
        }

        // Sort by value/weight ratio in descending order
        Arrays.sort(items, (a, b) ->
                Double.compare(b.ratio, a.ratio));

        double totalValue = 0.0;
        long remainingCapacity = capacity;

        for (Item item : items) {

            if (remainingCapacity == 0) {
                break;
            }

            // Take complete item
            if (item.weight <= remainingCapacity) {

                totalValue += item.value;
                remainingCapacity -= item.weight;

            } 
            // Take fractional part
            else {

                totalValue += item.ratio * remainingCapacity;
                remainingCapacity = 0;
            }
        }

        return totalValue;
    }
}